---
title: "CLL318 Del"
author: "Hongke Peng"
date: "2023-12-24"
output: html_document
---

#load library

```{r}
library(Seurat)
library(Biostrings)
library(GenomicRanges)
library(ggplot2)
library(Rsamtools)
library(rtracklayer)
library(GenomicAlignments)
library(msigdbr)
```

#find 6bp deletion
```{r}
fq <- list.files("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/demultiplex/")
fq <- fq[grep("fq.gz", fq)]
sample.ls <- sapply(strsplit(fq, "\\."), function(x){x[1]})
all.del <- data.frame()
all.wt <- data.frame()
for (i in sample.ls) {
  print(paste("now working on", i, "==================="))
  bamPath <- paste0("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/re_basecall/rachseq/fastq_polyA/filtered_bam/" , i, "_align2genome.bam")
  mygene <- GRanges(seqnames = "chr2", 
                    ranges = IRanges(197402103, 197402110),
                    strand = "-")
  x <- stackStringsFromBam(bamPath, param = mygene, use.names = T)
  x.del <- x[x == "C------T"]
  if(length(x.del) > 0){
  df.del <- data.frame(read_name = names(x.del))
  df.del$bc_umi <- sapply(strsplit(df.del$read_name, "#"), function(x){x[1]})
  df.del$bc <- sapply(strsplit(df.del$bc_umi, "_"), function(x){x[1]})
  df.del$umi <- sapply(strsplit(df.del$bc_umi, "_"), function(x){x[2]})
  
  #add 2023-08-03 to merge UMI
  df.del <- df.del[!(duplicated(df.del$bc_umi)), ]
  # =======
  
  stat.del <- data.frame(table(df.del$bc))
  colnames(stat.del) <- c("bc", "count")
  stat.del$bc <- paste(i, stat.del$bc, sep = "_")
  stat.del$patient <- i
  all.del <- rbind(all.del, stat.del)
  } else {
    print("There is no deletion in this patient")
  }
  
  x.wt <- x[x == "CGAACTTT"]
  if(length(x.wt) > 0){
  df.wt <- data.frame(read_name = names(x.wt))
  df.wt$bc_umi <- sapply(strsplit(df.wt$read_name, "#"), function(x){x[1]})
  df.wt$bc <- sapply(strsplit(df.wt$bc_umi, "_"), function(x){x[1]})
  df.wt$umi <- sapply(strsplit(df.wt$bc_umi, "_"), function(x){x[2]})
  
  #add 2023-0803 to merge UMI
  df.wt <- df.wt[!(duplicated(df.wt$bc_umi)), ]
  # =======
  
  stat.wt <- data.frame(table(df.wt$bc))
  colnames(stat.wt) <- c("bc", "count")
  stat.wt$bc <- paste(i, stat.wt$bc, sep = "_")
  stat.wt$patient <- i
  all.wt <- rbind(all.wt, stat.wt)
  } else {
    print("There is no WT in this patient")
  }
}
```

```{r}
sr <- readRDS("make_srt_object/sr_data.rds")
#b <- readRDS("../analysis/sf3b1_btk_step3_fig/cll_b.rds")
all.del$bc <- paste(all.del$bc, "1", sep = "-")
filter.all.del <- all.del[all.del$bc %in% colnames(sr), ]
filter.all.del <- filter.all.del[filter.all.del$count > 2, ] #rm background

all.wt$bc <- paste(all.wt$bc, "1", sep = "-")
filter.all.wt <- all.wt[all.wt$bc %in% colnames(sr), ]
filter.all.wt <- filter.all.wt[!(filter.all.wt$bc %in% all.del$bc), ]
filter.all.wt <- filter.all.wt[filter.all.wt$count > 5, ] #set a hard threshold for wt CLL
```

```{r, fig.width=5, fig.height=5}
sample.ls[!(sample.ls %in% unique(filter.all.del$patient))]
no.del <- data.frame(patient = sample.ls[!(sample.ls %in% unique(filter.all.del$patient))], 
                     nCell = 0)

sum.del <- as.data.frame(table(filter.all.del$patient))
colnames(sum.del) <- c("patient", "nCell")
sum.del <- rbind(sum.del, no.del)

pt_table <- read.csv("sf3b1_in_t_fig/patient_number.csv", header = F)
colnames(pt_table) <- c("sample", "alias")
sum.del$alias <- pt_table$alias[match(sum.del$patient, pt_table$sample)]
alias_order_adj <- c("CLL3-S", "CLL3-R", "CLL26-S", "CLL26-R", "CLL2-RB", "CLL2-R", "CLL5-R", "CLL5-RB", "CLL6-S", "CLL6-R", "CLL7-S", "CLL7-R", 
                     "CLL16-S", "CLL16-R", "CLL17-R", "CLL23-S", "CLL23-R", "CLL23-RB",
                     "HD11", "HD12", "HD14")
sum.del$alias <- factor(sum.del$alias, levels = alias_order_adj)

ggplot(sum.del, aes(x = alias, y = nCell)) + 
  geom_col() + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
ggsave("sf3b1_del_fig/cell_with_sf3b1_del_>2.pdf", width = 5, height = 5)
```

```{r, fig.width=5, fig.height=5}
bc_with_del <- filter.all.del$bc[filter.all.del$patient == "CLL318"]
sr$sf3b1_del <- "no_detection"
sr$sf3b1_del[colnames(sr) %in% bc_with_del] <- "deletion"

plot_df <- data.frame(sr@reductions$mnn_umap@cell.embeddings, 
                      deletion = sr$sf3b1_del)
plot_df_1 <- plot_df[plot_df$deletion == "deletion", ]
ggplot(plot_df) + 
  geom_point(aes(x = mnnumap_1, y = mnnumap_2), color = "grey", size = 0.1) +
  theme_bw() + 
  geom_point(data = plot_df_1, aes(x = mnnumap_1, y = mnnumap_2), color = "#BC3C2999", size = 1) +
  xlab("UMAP 1") + 
  ylab("UMAP 2")
ggsave("sf3b1_del_fig/umap_sf3b1_del_>2_red.pdf", width = 5, height = 5)
```

#collect
```{r, fig.width=10, fig.height=4}
filter.all.del
filter.all.wt

cll318.wt <- filter.all.wt[filter.all.wt$patient == "CLL318", ]
cll318.del <- filter.all.del[filter.all.del$patient == "CLL318", ]
cll318.merg <- merge(x = cll318.wt, y = cll318.del, by = "bc", all = TRUE)
cll318.merg$patient.x <- NULL
cll318.merg$patient.y <- NULL
colnames(cll318.merg) <- c("bc", "wt.count", "del.count")
cll318.merg[is.na(cll318.merg)] <- 0

sr$cell.type <- "CLL"
sr$cell.type[sr$refined.cluster == 0] <- "Monocytes"
sr$cell.type[sr$refined.cluster == 11] <- "B cells"
sr$cell.type[sr$refined.cluster %in% 12:16] <- "T cells"

pt <- subset(sr, orig.ident == "CLL318")
cll_name <- colnames(pt)[pt$cell.type == "CLL"]

cll318.merg$cluster <- "other_cluster"
cll318.merg$cluster[cll318.merg$bc %in% cll_name] <- "CLL_cluster"
p1 <- ggplot(cll318.merg, aes(x = wt.count, fill = cluster)) + 
  geom_bar(width = 1) + 
  xlim(c(0,50)) + 
  ylim(c(0,500)) + 
  theme_bw() +
  ggtitle("WT SF3B1 read counts") + 
  ylab("Cell counts") + 
  NoLegend() 
p2 <- ggplot(cll318.merg, aes(x = del.count, fill = cluster)) + 
  geom_bar(width = 1) + 
  xlim(c(0,50)) + 
  ylim(c(0,500)) + 
  ggtitle("6bp del SF3B1 read counts") + 
  ylab("Cell counts") + 
  theme_bw()
p1 + p2
```
#compare in CLL318 CLL cluster, 6bp del vs WT

```{r, fig.height=5, fig.width=5}
DimPlot(sr, group.by = "refined.cluster", label = T, cols = dittoColors(), order = F)

sr$cell.type <- "CLL"
sr$cell.type[sr$refined.cluster == 0] <- "Monocytes"
sr$cell.type[sr$refined.cluster == 8] <- "healthy_B"
sr$cell.type[sr$refined.cluster %in% 9:13] <- "T"

pt <- subset(sr, orig.ident == "CLL318")
plot_df <- data.frame(table(pt$cell.type, pt$sf3b1_del)[, 1])
colnames(plot_df) <- "nCell"
plot_df$cell.type <- rownames(plot_df)
ggplot(plot_df, aes(x = cell.type, y = nCell)) + 
  geom_col(width = 0.7) +
  theme_bw()
ggsave("sf3b1_del_in_CLL318_fig/nCell_in_cluster.pdf", width = 5, height = 5)
```

```{r}
cll318.del <- filter.all.del$bc[filter.all.del$patient == "CLL318"]
cll318.wt <- filter.all.wt$bc[filter.all.wt$patient == "CLL318"]

cll318.wt <- cll318.wt[!(cll318.wt %in% cll318.del)]

pt <- subset(sr, orig.ident == "CLL318")
pt$de_compare <- "other"
pt$de_compare[(colnames(pt) %in% cll318.del) & (pt$cell.type == "CLL") ] <- "SF3B1-Del"
pt$de_compare[(colnames(pt) %in% cll318.wt) & (pt$cell.type == "CLL")] <- "SF3B1-WT"
DefaultAssay(pt) <- "RNA"
marker <- FindMarkers(pt, group.by = "de_compare", ident.1 = "SF3B1-Del", ident.2 = "SF3B1-WT", test.use = "MAST", logfc.threshold = 0, min.diff.pct = 0)
marker$gene <- rownames(marker)
```

```{r,fig.width=6, fig.height=7}
EnhancedVolcano::EnhancedVolcano(marker,
                x="avg_log2FC",y="p_val_adj",
                lab=rownames(marker),
                #xlim=c(-10,10),
                #ylim=c(0,50),
                title = "CLL318-Del vs CLL318-WT",
                subtitle = paste0('FDR cutoff = 0.01', "  logFC cutoff = 0.25"),
                labSize = 4,
                legendPosition = "top",
                legendLabSize = 12,
                legendIconSize = 4.0,
                raster = T,
                max.overlaps = 15,
                boxedLabels = T,
                drawConnectors = T,
                pCutoff = 0.05,
                FCcutoff = 0.25)
ggsave("sf3b1_del_fig/CLL318_mut_vs_wt.pdf", width = 6, height = 7)
```

```{r,fig.width=3, fig.height=5}
df <- table(pt$de_compare) %>% as.data.frame()
colnames(df) <- c("group", "nCell")
ggplot(df, aes(x = group, y = nCell)) + 
  geom_col(width = 0.6) + 
  theme_bw() + 
  ylab("Cell counts") + 
  ggtitle("CLL318 CLL cells involved in DE analysis")
ggsave("sf3b1_del_in_CLL318_fig/cll318_cell_counts_in_de_5reads.pdf", width = 3, height = 5)
```


#all 6bp-del vs all wt
```{r}
#cll <- subset(sr, (cell.type == "CLL") & (orig.ident %in% c("CLL141v", "CLL141i", "CLL153", "CLL63v", "CLL63i", "CLL232", "CLL152", "CLL318", "CLL155", "CLL170v", "CLL170i", "CLL175")) )
sr$cell.type <- "CLL"
sr$cell.type[sr$refined.cluster == 0] <- "Monocytes"
sr$cell.type[sr$refined.cluster == 11] <- "healthy_B"
sr$cell.type[sr$refined.cluster %in% 12:16] <- "T"
cll <- subset(sr, (cell.type == "CLL") & (orig.ident %in% c("CLL141v", "CLL63v", "CLL232", "CLL152", "CLL318", "CLL170v", "CLL175")))
DimPlot(cll, group.by = "refined.cluster", label = T, cols = dittoColors(), order = F)

del <- filter.all.del$bc
wt <- filter.all.wt$bc[!(filter.all.wt$bc %in% del)]

cll$de_compare <- "other"
cll$de_compare[colnames(cll) %in% del] <- "SF3B1-Del"
cll$de_compare[colnames(cll) %in% wt] <- "SF3B1-WT"
cll <- subset(cll, de_compare != "other")
DefaultAssay(cll) <- "RNA"
marker <- FindMarkers(cll, group.by = "de_compare", ident.1 = "SF3B1-Del", ident.2 = "SF3B1-WT", test.use = "MAST", logfc.threshold = 0, min.diff.pct = 0)

marker$gene <- rownames(marker)
write.csv(marker, "sf3b1_del_fig/sf3b1_del_vs_all_wt.csv")
#marker <- read.csv("sf3b1_del_fig/sf3b1_del_vs_all_wt.csv")
selected_marker <- marker[!grepl("^IG", marker$gene), ]
```

```{r,fig.width=7, fig.height=8}
library(EnhancedVolcano)
EnhancedVolcano(selected_marker,
                x="avg_log2FC",y="p_val_adj",
                lab=rownames(selected_marker),
                #xlim=c(-10,10),
                #ylim=c(0,50),
                title = "SF3B1-Del vs SF3B1-WT",
                subtitle = paste0('FDR cutoff = 0.05', "  logFC cutoff = 0.5"),
                labSize = 5,
                legendPosition = "top",
                legendLabSize = 12,
                legendIconSize = 4.0,
                max.overlaps = 25,
                boxedLabels = T,
                raster = T,
                drawConnectors = T,
                pCutoff = 0.05,
                FCcutoff = 0.5)
ggsave("sf3b1_del_fig/6bp_vs_all_wt.pdf", width = 7, height = 8)
```