---
title: "Make Seurat Object"
author: "Hongke Peng"
date: "2023-12-11"
output: html_document
---


#load library
```{r}
library(SingleCellExperiment)
library(dittoSeq)
library(DropletUtils)
library(dplyr)
library(ggplot2)
library(Seurat)
library(ggpointdensity)
library(harmony)
library(transPlotR)
library(biomaRt)
library(Seurat)
library(SingleCellExperiment)
```


#Make single cell object for downstream analysis
pass reads with TSO and polyA tails to multi-sample pipeline
```{r}
sce <- readRDS("/vast/scratch/users/peng.h/rachseq/fastq_polyA/flames_out/sces.rds")

#Add prefix "CLLxxx_" to the cell barcodes
for (i in names(sce)) {
  x <- sce[[i]]
  colnames(x) <- paste(i, colnames(x), sep = "_")
  sce[[i]] <- x
}
```


##targets vs background
###logCount of targets and background
```{r, fig.width=5, fig.height=3}
x <- sce[[names(sce)[1]]]
all_df <- rowData(x) %>% as.data.frame()
all_df$count <- rowSums(counts(x))
all_df$nCell <- rowSums(counts(x) > 0)

for (i in names(sce)[2:length(sce)]) {
  x <- sce[[i]]
  df <- rowData(x) %>% as.data.frame()
  df$count <- rowSums(counts(x))
  df$nCell <- rowSums(counts(x) > 0)
  all_df <- rbind(all_df, df)
}

#make a df that have gene_id and its corresponding FSM
geneid_FSM <- all_df[!duplicated(all_df$FSM_match), ]

#marge transcripts by FSM
df <- all_df
df$transcript_id <- NULL
df$gene_id <- NULL
df <- df %>% group_by(FSM_match) %>% summarise(count = sum(count),
                                               nCell = sum(nCell))
df$logCount <- log10(df$count)
df$lognCell <- log10(df$nCell)

df$gene_id <- geneid_FSM$gene_id[match(df$FSM_match, geneid_FSM$FSM_match)]
df$symbol <- "background"
df$symbol[df$gene_id == "ENSG00000115524.16"] <- "SF3B1"
df$symbol[df$gene_id == "ENSG00000010671.15"] <- "BTK"
df$symbol <- factor(df$symbol, levels = c("SF3B1", "BTK", "background"))


p1 <- ggplot(df, aes(x = symbol, y = logCount, color = symbol)) +
  geom_jitter(aes(colour = symbol), position = position_jitter(0.1), alpha = 1, size = 0.5) +
  geom_violin(aes(fill = symbol, colour = symbol), alpha = 0.6, scale = "area", trim = F, width = 0.5) +
  scale_fill_manual(values = c("#E47250",  "#EBB261", "#5A4A6F")) +
  scale_colour_manual(values = c("#E47250",  "#EBB261", "#5A4A6F")) +
  ylab("log10 UMIs") +
  xlab(NULL) +
  #ylim(c(0, 5000)) + 
  theme_classic() + 
  theme(axis.text = element_text(size = 12))
p1
ggsave("make_srt_object/QC_target_bg_UMIs.pdf", width = 5, height = 3)
```


###log nCell of targets and background
```{r, fig.width=5, fig.height=3}
p2 <- ggplot(df, aes(x = symbol, y = lognCell, color = symbol)) +
  geom_violin(aes(fill = symbol, colour = symbol), alpha = 0.6, scale = "area", trim = F, width = 0.5) +
  geom_jitter(aes(colour = symbol), position = position_jitter(0.1), alpha = 1, size = 0.3) +
  scale_fill_manual(values = c("#E47250",  "#EBB261", "#5A4A6F")) +
  scale_colour_manual(values = c("#E47250",  "#EBB261", "#5A4A6F")) +
  ylab("log10 cell counts") + 
  xlab(NULL) + 
  theme_classic() + 
  theme(axis.text = element_text(size = 12))
p2
ggsave("make_srt_object/QC_target_bg_nCell.pdf", width = 5, height = 3)
```

###find the top off-target tr
```{r}
library(ggrepel)

#give symbol to genes
mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
genes <- df$gene_id
genes <- sapply(strsplit(genes, "\\."), function(x){x[1]})
G_list <- getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id", "hgnc_symbol", "description"), values = genes, mart = mart)

df$gene_id <- sapply(strsplit(df$gene_id , "\\."), function(x){x[1]})
sum(df$gene_id %in% G_list$ensembl_gene_id) #4367, which means almost all ensembl in map_fsm can be found in G list
df$hgnc_symbol <- G_list$hgnc_symbol[match(df$gene_id, G_list$ensembl_gene_id)]

gb_df <- df[df$symbol == "background", ]
gb_df$axisX <- runif(nrow(gb_df), -1, 1)
gb_df <- gb_df[order(gb_df$logCount, decreasing = T), ]
y <- gb_df[1:10, ]

p3 <- ggplot(gb_df, aes(x = axisX, y = logCount)) + 
  geom_point(size = 0.3, color = "grey50") + theme_bw() + 
  xlab(NULL) + ylab("log10 read count") + ggtitle("Top 10 off-target transcripts") + 
  ylim(c(0, 1.2 * max(y$logCount))) + 
  geom_point(data = y, aes(x = axisX, y = logCount), size = 1.5, color = "red") + 
  geom_label_repel(data = y, aes(x = axisX, y = logCount, label = hgnc_symbol), direction = "both", arrow = arrow(length = unit(0.02, "npc")), seed = 1, max.overlaps = 20)
```

```{r, fig.width=14, fig.height=4}
p <- p1 +  p2 + p3
p
ggsave(plot = p, filename = "make_srt_object/QC_target_vs_bg.pdf", width = 14, height = 5)
```

#make srt
#load data
```{r, fig.width=8, fig.height=4}
folder <- list.files("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/fastq/")
folder <- folder[!grepl("csv", folder)]
dir_list <- c()
for (i in folder) {
  dir_list[i] <- paste("/stornext/Genomics/data/CLL_venetoclax/workspace/T_under_VEN/Aim1/cr_output/", i, "/filtered_feature_bc_matrix80/", sep = "")
}

data <- Read10X(dir_list)
srt <- CreateSeuratObject(counts = data$`Gene Expression`, project = "SF3B1", min.cells = 3, min.features = 200)
srt$project <- "LR"
srt[["percent.mt"]] <- PercentageFeatureSet(srt, pattern = "^MT-")

VlnPlot(srt, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = 0, group.by = "project")
```


```{r, fig.width=5, fig.height=5}
flames.ls <- readRDS("/vast/scratch/users/peng.h/rachseq/fastq_polyA/flames_out/sces.rds")
#Add prefix "CLLxxx_" to the cell barcodes
for (i in names(flames.ls)) {
  x <- flames.ls[[i]]
  colnames(x) <- paste(i, colnames(x), sep = "_")
  colnames(x) <- paste(colnames(x), "1", sep = "-")
  flames.ls[[i]] <- x
}

#find cell barcodes detected in long-read
bc <- c()
for (i in names(flames.ls)) {
  x <- flames.ls[[i]]
  bc <- c(bc, colnames(x))
}
#select cells in short read data
keep_cell <- bc[bc %in% colnames(srt)] #compared to short-read data, we loss 8827 cells
sr <- srt[, keep_cell]
```


#make UMAP
```{r}
DefaultAssay(sr) <- "RNA"
sr <- NormalizeData(sr)
sr <- FindVariableFeatures(sr, nfeatures = 3000)
sr <- ScaleData(sr)
sr <- RunPCA(sr, reduction.name = "uncorr_pca", reduction.key="uncorr_pca_")
ElbowPlot(object = sr, ndims=50, reduction="uncorr_pca")
```

```{R,fig.width = 8,fig.height = 6}
dims.to.use <- 30
sr <- RunUMAP(sr, reduction = "uncorr_pca", dims = 1:dims.to.use, reduction.name= "uncorr_umap", reduction.key = "uncorr_umap_")
DimPlot(object = sr, reduction = "uncorr_umap", group.by = "orig.ident", cols=c(ggsci::pal_npg(alpha=0.7)(8), ggsci::pal_nejm(alpha=0.7)(8), ggsci::pal_jama(alpha = 0.7)(8)), order=T)
```
##run harmony
```{r}
theta.to.use <- 0
sr=RunHarmony(sr,
               group.by.vars="orig.ident",
               reduction="uncorr_pca",
               theta=theta.to.use,
               plot_convergence=TRUE,
               reduction.save="harmony_pca",
               kmeans_init_nstart=1000,
               kmeans_init_iter_max=500000)
```

```{r, fig.width=7, fig.height=5}
sr <- RunUMAP(sr ,reduction = "harmony_pca", dims=1:dims.to.use, reduction.name = "harmony_umap", reduction.key = "harmony_umap_")
DimPlot(object = sr, reduction = "harmony_umap", group.by = "orig.ident", cols=c(ggsci::pal_npg(alpha=0.7)(8), ggsci::pal_nejm(alpha=0.7)(8), ggsci::pal_jama(alpha = 0.7)(8)), order=T)
```

```{r, fig.width=15, fig.height=15}
#add stage to sr
sr$stage <- "relapsed"
DimPlot(object = sr, reduction = "harmony_umap", group.by = "orig.ident", cols=c(ggsci::pal_npg(alpha=0.7)(8), ggsci::pal_nejm(alpha=0.7)(8), ggsci::pal_jama(alpha = 0.7)(8)), order=T, split.by = "orig.ident", ncol = 5)
```

##run fastMNN
```{r}
#load package
library(Seurat)
library(batchelor)
library(BiocNeighbors)
library(BiocSingular)
library(BiocParallel)
bpp=MulticoreParam(10)

sr <- NormalizeData(sr)
sr <- FindVariableFeatures(sr, nfeatures = 3000)
hvg <- VariableFeatures(sr)

correction.order=list(list("CLL156","CLL5309","CLL5602","CLL5305","CLL5306","CLL108"),#screening
                      list("CLL141v","CLL141i","CLL153","CLL63i","CLL63v","CLL232","CLL152","CLL318","CLL155","CLL170i", "CLL170v", "CLL175"),#VEN relapsed
                      list("HD11","HD12","HD14") #HD samples
)

set.seed(2022)
print("start run mnn -----------------------------")
corr.f=fastMNN(sr@assays$RNA@data,
               batch=factor(sr$orig.ident),
               k=30,
               d=50,
               deferred=T,
               auto.merge=T,
               merge.order = correction.order, 
               subset.row=hvg,
               BSPARAM=RandomParam(),
               BPPARAM=bpp)

#saveRDS(corr.f,"/stornext/Genomics/data/CLL_venetoclax/workspace/SF3B1_mutation/analysis/Make_srt_step2_fig/mnn.rds")

#add MNN corrected pca to sr
mnn_pca <- reducedDim(corr.f, "corrected")
colnames(mnn_pca) <- paste("PCA", 1:50, sep = "_")
sr[["mnn_pca"]] <- Seurat::CreateDimReducObject(embeddings = mnn_pca)
```


###make umap
```{r, fig.width=6, fig.height=6}
#make umap
dims.to.use <- 30
sr <- RunUMAP(sr ,reduction = "mnn_pca", dims=1:dims.to.use, reduction.name = "mnn_umap", reduction.key = "mnn_umap_")

#add annotation for samples
sr$stage <- "relapsed"
sr$stage[sr$orig.ident %in% c("CLL156","CLL5309","CLL5602","CLL5305","CLL5306","CLL108")] <- "screening"
sr$stage[sr$orig.ident %in% c("HD11","HD12","HD14")] <- "HD"
sr$stage <- factor(sr$stage, levels = c("screening", "relapsed", "HD"))

DimPlot(object = sr, reduction = "mnn_umap", group.by = "stage", order = F, cols = c("#84B1D5", "#D09DCB", "#6BB8D2"))
```

###find cluster
```{r, fig.width=6, fig.height=6}
sr <- FindNeighbors(sr, reduction = "mnn_pca", dims = 1:dims.to.use, graph.name = "mnn")
sr <- FindClusters(sr, graph.name = "mnn", resolution = 0.8)
DimPlot(object = sr, group.by = "mnn_res.1",
        reduction = "mnn_umap", 
        order = F, 
        cols = c(ggsci::pal_npg(alpha=0.7)(8), ggsci::pal_nejm(alpha=0.7)(8), ggsci::pal_jama(alpha = 0.7)(8)), 
        label = T)
```

```{r, fig.width=15 ,fig.height=15}
table(sr$mnn_res.1)
DimPlot(object = sr, reduction = "mnn_umap", order = F, split.by = "orig.ident", ncol = 5,
        cols = c(ggsci::pal_npg(alpha=0.7)(8), ggsci::pal_nejm(alpha=0.7)(8), ggsci::pal_jama(alpha = 0.7)(8)), 
        label = T)
```


###short-read data QC
```{r, fig.width=15, fig.height=8}
VlnPlot(sr, features = c("nCount_RNA", "nFeature_RNA", "percent.mt", "CD19", "MS4A1", "TCL1A", "BTK", "CD3E", "GNLY", "NKG7", "SF3B1", "CD14", "NCAM1"), pt.size = 0, cols = c(ggsci::pal_npg(alpha=0.7)(8), ggsci::pal_nejm(alpha=0.7)(8), ggsci::pal_jama(alpha = 0.7)(8)))
```

###remove cluster
rm cluster 10: doublets
rm cluster 9, 11
rm cluster 18, 19
I try to retains as much as cell we can. 
```{r, fig.width=6 ,fig.height=6}
#saveRDS(sr, "make_srt_object/sr_data.rds")
#sr <- subset(sr, seurat_clusters %in% c(0:8, 12:17))
DimPlot(object = sr, reduction = "mnn_umap", order = F, 
        cols = c(ggsci::pal_npg(alpha=0.7)(8), ggsci::pal_nejm(alpha=0.7)(8), ggsci::pal_jama(alpha = 0.7)(8)), 
        label = T)
dittoDimPlot(sr, var = "mnn_res.1", reduction.use = "mnn_umap", do.label = T, do.raster = T, size = 0.1, 
             color.panel = c(ggsci::pal_npg(alpha=0.5)(7), ggsci::pal_nejm(alpha=0.5)(7), ggsci::pal_jama(alpha = 0.5)(7)))
#loop back
```
###make good umap figure
```{r, fig.width=5, fig.height=5}
sr$refined.cluster <- 0
sr$refined.cluster[sr$mnn_res.1 == 1] <- 1
sr$refined.cluster[sr$mnn_res.1 == 0] <- 2
sr$refined.cluster[sr$mnn_res.1 == 15] <- 3
sr$refined.cluster[sr$mnn_res.1 == 10] <- 4
sr$refined.cluster[sr$mnn_res.1 == 13] <- 5
sr$refined.cluster[sr$mnn_res.1 == 12] <- 6
sr$refined.cluster[sr$mnn_res.1 == 6] <- 7
sr$refined.cluster[sr$mnn_res.1 == 5] <- 8
sr$refined.cluster[sr$mnn_res.1 == 2] <- 9
sr$refined.cluster[sr$mnn_res.1 == 8] <- 10
sr$refined.cluster[sr$mnn_res.1 == 11] <- 11
sr$refined.cluster[sr$mnn_res.1 == 14] <- 12
sr$refined.cluster[sr$mnn_res.1 == 7] <- 13
sr$refined.cluster[sr$mnn_res.1 == 3] <- 14
sr$refined.cluster[sr$mnn_res.1 == 9] <- 15
sr$refined.cluster[sr$mnn_res.1 == 4] <- 16

colset <- c("#A9A9A9", 
            "#F055A3", "#A56AAC", "#D99ACE", "#DCBFDD", "#77B3DA", "#4CBBD5", "#A5D0F0", "#528DD9", "#7876B1B2","#4DBBD5B2","#00A1D5B2",
            "#9BCA3C", "#91D1C2", "#5FB14E", "#CADB73", "#3B897C")
  
DimPlot(object = sr, reduction = "mnn_umap", order = F, group.by = "refined.cluster", 
        cols = colset, raster = T, pt.size = 0.3,
        label = T, label.box = T, label.color = "white", label.size = 3)

ggsave(filename = "make_srt_object/sr_umap.pdf", width = 5, height = 5)
```

```{r}
saveRDS(sr, "make_srt_object/sr_data.rds")
```


### make gene express plot
```{r, fig.width=10, fig.height=2.5}
p1 <- dittoDimPlot(sr, var = "stage", reduction.use = "mnn_umap", do.label = F, do.raster = F, size = 0.0001, order = "decreasing")
FeaturePlot(sr, features = c("SF3B1", "BTK"), max.cutoff = "q98", min.cutoff = "q2", reduction = "mnn_umap", order = T)
VlnPlot(sr, features = c("SF3B1", "BTK"), pt.size = 0, cols = colset, group.by = "refined.cluster")
```




#make lr matrix
```{r}
sce <- readRDS("/vast/scratch/users/peng.h/rachseq/fastq_polyA/flames_out/sces.rds")

library(scMerge)
#Add prefix "CLLxxx_" to the cell barcodes
#select cells that have sr data
for (i in names(sce)) {
  x <- sce[[i]]
  colnames(x) <- paste(i, colnames(x), sep = "_")
  colnames(x) <- paste(colnames(x), "1", sep = "-")
  keep_cell <- colnames(x)[colnames(x) %in% colnames(sr)]
  x <- x[, keep_cell]
  #put the gene id to the start of the rowname
  rownames(x) <- paste(rowData(x)$gene_id, rownames(x), sep = "_")
  sce[[i]] <- x
}
#combining all sce
combined.sce <- sce_cbind(sce_list = sce, 
                  exprs = "counts", 
                  method = "union", 
                  cut_off_batch = 0, 
                  cut_off_overall = 0)
combined.sce <- combined.sce[, colnames(sr)]

#all lr matrix to sr data
sr[["tr"]] <- CreateAssayObject(counts = counts(combined.sce))
#saveRDS(sr, "Make_srt_step2_fig/sr_data.rds")
```

#correlation between SF3B1 and BTK
```{r}
sr <- readRDS("make_srt_object/sr_data.rds")

sr$refined.cluster <- factor(sr$refined.cluster)
dittoDimPlot(sr, var = "refined.cluster", reduction.use = "mnn_umap", do.label = T, do.raster = T, size = 0.2, labels.size = 5,  
                   #color.panel = colset
                   )
```

```{r}
cor.df <- data.frame(barcode = colnames(sr), 
                     sr_sf3b1 = sr@assays$RNA@counts["SF3B1",],
                     sr_btk = sr@assays$RNA@counts["BTK",])

tr_mtx <- GetAssayData(sr, assay = "tr", slot = "count")
#add in lr sf3b1 data
sf3b1_mtx <- tr_mtx[grepl("ENSG00000115524.16", rownames(tr_mtx)), ]
cor.df$lr_sf3b1 <- colSums(sf3b1_mtx)
#add in lr BTK data
btk_mtx <- tr_mtx[grepl("ENSG00000010671.15", rownames(tr_mtx)), ]
cor.df$lr_btk <- colSums(btk_mtx)  

cor.df$sample <- sapply(strsplit(cor.df$barcode, "_"), function(x){x[1]})

plot.df <- cor.df %>% group_by(sample) %>% summarise(sr_sf3b1 = sum(sr_sf3b1), 
                                                     lr_sf3b1 = sum(lr_sf3b1), 
                                                     sr_btk = sum(sr_btk), 
                                                     lr_btk = sum(lr_btk))
#cor by sample
ggplot(plot.df, aes(x = sr_sf3b1, y = lr_sf3b1)) + 
  geom_point() + 
  theme_classic() + 
  theme(axis.text = element_text(size = 12), axis.title = element_text(size = 15)) +
  scale_x_continuous(limits = c(0, 25000),expand = c(0, 0)) +
  scale_y_continuous(limits = c(0, 100000),expand = c(0, 0)) +
  ggtitle("SF3B1, correlation of lr and sr")

ggplot(plot.df, aes(x = sr_btk, y = lr_btk)) + 
  geom_point() + 
  theme_classic() + 
  theme(axis.text = element_text(size = 12), axis.title = element_text(size = 15)) +
  scale_x_continuous(limits = c(0, 6000),expand = c(0,200)) +
  scale_y_continuous(limits = c(0, 50000),expand = c(0,200)) +
  ggtitle("BTK, correlation of lr and sr")
```

```{r, fig.width=5,fig.height=5}
ncell.df <- table(cor.df$sample) %>% as.data.frame()
plot.df$nCell <- ncell.df$Freq

sf3b1 <- data.frame(sample = plot.df$sample, 
                    sr_sf3b1 = plot.df$sr_sf3b1 /plot.df$nCell, 
                    lr_sf3b1 = plot.df$lr_sf3b1 /plot.df$nCell)

btk <- data.frame(sample = plot.df$sample, 
                    sr_btk = plot.df$sr_btk /plot.df$nCell, 
                    lr_btk = plot.df$lr_btk /plot.df$nCell)

ggplot(sf3b1, aes(x = sr_sf3b1, y = lr_sf3b1)) + 
  geom_point(size = 3, color = "#E47250", alpha = 0.8) + 
  geom_point(data = btk, aes(x = sr_btk, y = lr_btk), size = 3, color = "#EBB261", alpha = 0.8) + 
  theme_classic() + 
  theme(axis.text = element_text(size = 12), axis.title = element_text(size = 15)) +
  #geom_smooth(method = "lm", se=FALSE, color="grey80", formula = y ~ x, linetype = "dashed") + 
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "grey60") + 
  xlab("SF3B1 UMIs in short-read data") + 
  ylab("SF3B1 UMIs in long-read data") +
  scale_x_continuous(limits = c(0, 13),expand = c(0,0)) +
  scale_y_continuous(limits = c(0, 13),expand = c(0,0)) +
  ggtitle("SF3B1, correlation of lr and sr")

ggsave("make_srt_object/lr_sr_correlation.pdf", width = 5, height = 5)
```


#make UMAP
```{r}
sr <- readRDS("Make_srt_step2_fig/sr_data.rds")

colset <- c("#A9A9A9", 
            "#F055A3", "#A56AAC", "#D99ACE", "#DCBFDD", "#77B3DA", "#4CBBD5", "#A5D0F0", "#528DD9", 
            "#9BCA3C", "#91D1C2", "#5FB14E", "#CADB73", "#3B897C")

sr$refined.cluster <- factor(sr$refined.cluster)
p1 <- dittoDimPlot(sr, var = "refined.cluster", reduction.use = "mnn_umap", do.label = T, do.raster = T, size = 0.2, labels.size = 5,  
                   color.panel = colset)
p1
ggsave("Make_srt_step2_fig/sr_umap.pdf", width = 6, height = 5)
```


```{r}
dittoDimPlot(sr, var = "stage", reduction.use = "mnn_umap", do.label = F, do.raster = T, size = 0.01, labels.size = 5)
DimPlot(object = sr, reduction = "mnn_umap", order = F, group.by = "stage", 
        cols = c("#E18727CC", "#BC3C29CC", "#0072B5CC"), 
        #cols = colset, 
        raster = T, 
        pt.size = 0.01,
        label = F, label.box = F, label.color = "white", label.size = 3)
ggsave("Make_srt_step2_fig/sr_umap_col_by_stage.pdf", width = 6.5, height = 5)
```

## short read expression of BTK and SF3B1
```{r, fig.width=10, fig.height=5}
DefaultAssay(sr) <- "RNA"
sr <- NormalizeData(sr)
sr <- FindVariableFeatures(sr, nfeatures = 3000)

VlnPlot(sr, assay = "RNA", group.by = "refined.cluster", 
        features = c("CD14", "CD19", "CD3E"), cols = colset, 
        pt.size = 0, ncol = 1)
ggsave("make_srt_object/vln.pdf", width = 5, height = 5.5)

FeaturePlot(sr, reduction = "mnn_umap", features = c("CD19", "CD3E"), min.cutoff = "q2", max.cutoff = "q98", order = T, pt.size = 0.1)
ggsave("Make_srt_step2_fig/b_t_marker.pdf", width = 10, height = 5)

ploat.ls <- list()
for (i in c("SF3B1", "BTK")) {
  ploat.ls[[i]] <- FeaturePlot(sr, reduction = "mnn_umap", features = i, min.cutoff = "q2", max.cutoff = "q98", order = T, pt.size = 0.1) + 
    xlab(NULL) + ylab(NULL) +
    theme(axis.text = element_blank())
}

pdf("make_srt_object/sf3b1_btk.pdf", width = 10/1.5, height = 5/1.5)
gridExtra::grid.arrange(grobs = ploat.ls, ncol = 2)
dev.off()
ggsave("Make_srt_step2_fig/sf3b1_btk.pdf", width = 10/1.5, height = 5/1.5)
```

## long read expression of BTK and SF3B1

```{r}
tr_mtx <- GetAssayData(sr, assay = "tr", slot = "count")
tr_mtx <- as.data.frame(tr_mtx)
tr_mtx$ensembl <- sapply(base::strsplit(rownames(tr_mtx), "-"), function(x){x[1]})
tr_mtx$gene <- "offtarget"
tr_mtx$gene[tr_mtx$ensembl == "ENSG00000115524.16"] <- "SF3B1"
tr_mtx$gene[tr_mtx$ensembl == "ENSG00000010671.15"] <- "BTK"

tr_mtx$ensembl <- NULL

#calculate SF3B1 expression in lr data
sf3b1_df <- tr_mtx[tr_mtx$gene == "SF3B1", ]
sf3b1_df$gene <- NULL
sf3b1_df <- data.frame(SF3B1 = colSums(sf3b1_df))

#calculate BTK expression in lr data
btk_df <- tr_mtx[tr_mtx$gene == "BTK", ]
btk_df$gene <- NULL
btk_df <- data.frame(BTK = colSums(btk_df))

#make count mtx
combined_mtx <- cbind(sf3b1_df, btk_df)
combined_mtx <- t(combined_mtx)

#add new assay to sr
sr[["combined_tr"]] <- CreateAssayObject(counts = combined_mtx)
DefaultAssay(sr) <- "combined_tr"
sr <- NormalizeData(sr, normalization.method = "CLR")
sr <- ScaleData(sr)

ploat.ls <- list()
for (i in c("SF3B1", "BTK")) {
  ploat.ls[[i]] <- FeaturePlot(sr, reduction = "mnn_umap", features = i, min.cutoff = "q2", max.cutoff = "q98", order = T, pt.size = 0.1, raster = T,
                               cols = c("lightgrey", "darkgreen")) + 
                   xlab(NULL) + ylab(NULL) +
                   theme(axis.text = element_blank())
}

pdf("make_srt_object/sf3b1_btk_lr.pdf", width = 10/1.5, height = 5/1.5)
gridExtra::grid.arrange(grobs = ploat.ls, ncol = 2)
dev.off()
```





```{r}
saveRDS(sr, "make_srt_object/sr_data.rds")
rm(sr)
```

