---
title: "Remove atefacts"
author: "Hongke Peng"
date: "2023-12-11"
output: html_document
---

```{r}
library(Biostrings)
library(GenomicAlignments)
library(scatterpie)
library(ggplot2)
library(Rsamtools)
library(rtracklayer)
```

```{r}
mismatch_plot <- list()
fq <- list.files("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/demultiplex/")
fq <- fq[grep("fq.gz", fq)]
sample.ls <- sapply(strsplit(fq, "\\."), function(x){x[1]})

for(i in sample.ls){
print(paste("now working on", i, "================================="))
bamPath <- paste0("//vast/scratch/users/peng.h/rachseq/fastq_polyA/flames_out/", i, "_align2genome.bam")

mygene <- GRanges(seqnames = "chr2", 
                  ranges = IRanges(197405131, 197405135),
                  strand = "-")
x <- stackStringsFromBam(bamPath, param = mygene, use.names = T)
table(x)
df <- data.frame(read_name = names(as.character(x)), pos = as.character(x))
df$pos_1 <- substring(df$pos, 1, 1) == "T"
df$pos_2 <- substring(df$pos, 3, 3) == "T"
df$pos_3 <- substring(df$pos, 5, 5) == "T"
df$mis_match_score <- rowSums(df[, c("pos_1", "pos_2", "pos_3")])

#visualize mis-matches
plot_df <- data.frame(table(df$mis_match_score))
colnames(plot_df) <- c("mismatch", "read number")
mismatch_plot[[i]] <- ggplot(plot_df, aes(x = mismatch, y = `read number`)) + 
  geom_col(width = 0.7) + 
  theme_bw() + 
  xlab("C to T mis-matches in Chr2:197405131:197405135") + 
  theme(axis.text = element_text(size = 12), 
        axis.title = element_text(size = 12))

#select the row with mis-match score > 0 
df <- df[df$mis_match_score > 0, ] # all the reads in df are artifacts

#make filtered bam file
bamPath <- paste0("/vast/scratch/users/peng.h/rachseq/fastq_polyA/flames_out/", i, "_align2genome.bam")
#select SF3B1 and BTK
mygene <- GRanges(seqnames = c("chr2", "chrX"), 
                    ranges = c(IRanges(197380000, 197440000), IRanges(101345000, 101394000)),
                    strand = "-")
mybam <- readGAlignments(bamPath, 
                         use.names = T, 
                         param=ScanBamParam(which = mygene, what=c("seq","mapq","flag","isize","qual", "pos")))

keep_read <- !(names(mybam) %in% df$read_name)
mybam <- mybam[keep_read]
outPath <- paste0("/vast/scratch/users/peng.h/rachseq/fastq_polyA/filtered_bam/", i, "_align2genome.bam")
export(mybam, con = BamFile(outPath), format = "bam")

#make filtered fastq file
y <- Biostrings::readDNAStringSet(paste("/vast/scratch/users/peng.h/rachseq/fastq_polyA/", i, ".fq.gz", sep = ""), format = "fastq")
keep_read <- !(names(y) %in% df$read_name)
y <- y[keep_read]
### save reads with polyA
writeXStringSet(y, paste0("/vast/scratch/users/peng.h/rachseq/fastq_polyA/filtered_fastq/", i, ".fq.gz", sep = ""), format = "fastq")
}
```