---
title: "read coverage"
author: "Hongke Peng"
date: "2023-02-11"
output: html_document
---

```{r}
library(data.table)
library(Gviz)
library(Rsamtools)
library(rtracklayer)
library(GenomicAlignments)
library(Rsamtools)
```

```{r}
bamPath <- "/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/re_basecall/rachseq/fastq_polyA/flames_out/CLL141i_align2genome.bam"
bamPath <- "/stornext/Genomics/data/CLL_venetoclax/FLTseq/cll141/cll141_isoform_outs/align2genome.bam"
bamPath <- "~/public_home/Rachel/all_fastq/CLL141-CLL-cells_S8_Rsubread.sorted.bam"

mygene <- GRanges(seqnames = c("chr2"), 
                  ranges = c(IRanges(197380000, 197440000)),
                  strand = "-")

mybam <- readGAlignments(bamPath, 
                         use.names = T, 
                         param=ScanBamParam(which = mygene, what=c("seq","mapq","flag","isize","qual", "pos")))

export(mybam, 
       con = BamFile("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/re_basecall/rachseq/fastq_polyA/flames_out/CLL141i_bedcoverage/short_read.bam"),
       format = "bam")

#extract all chr2, since artifacts are in sf3b1
is_on_chr2 <- seqnames(mybam) == "chr2"
table(is_on_chr2)
sf3b1.read <- mybam[is_on_chr2]
```



```{r, fig.width=15, fig.height=5}
#bedgraph_dt <- fread("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/filtered_polyA/flames_out/CLL141i_bedcoverage/CLL141i_RaCHseq.bedgraph", 
#                     skip = 197388500, nrows = 197435079 - 197388500, 
#                    col.names = c('chromosome', 'start', 'value'))

#thechr <- "chr2"
#st <- 197388500
#en <- 197435079

#bedgraph_dt_one_chr <- bedgraph_dt[chromosome == thechr]
#bedgraph_dt_one_chr$end <- bedgraph_dt_one_chr$start
#dtrack <- DataTrack(
#    range = bedgraph_dt_one_chr,
#   type = "a",
#   genome = 'hg19',
#    name = "Sequencing Depth"
#)

thechr <- "chr2"
st <- 197388500
en <- 197435079

#itrack <- IdeogramTrack(genome = "hg19", chromosome = "chr2")
gtrack <- GenomeAxisTrack()

altrack <- AlignmentsTrack("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/re_basecall/rachseq/fastq_polyA/flames_out/CLL141i_align2genome.bam", 
                           isPaired = F, chromosome = "chr2", start = 197388500, end = 197435079,
                           col.mates = "deeppink")

altrack1 <- AlignmentsTrack("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/FLTseq_data/filtered_polyA/flames_out/cll141_align2genome.bam", 
                           isPaired = F, chromosome = "chr2", start = 197388500, end = 197435079,
                           col.mates = "deeppink")

altrack2 <- AlignmentsTrack("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/filtered_polyA/flames_out/CLL141i_bedcoverage/short_read.bam", 
                           isPaired = F, chromosome = "chr2", start = 197388500, end = 197435079,
                           col.mates = "deeppink")


library(BSgenome.Hsapiens.UCSC.hg19)
#strack <- SequenceTrack(
#    Hsapiens,
#    chromosome = thechr, from = en, to = st,
#    cex=0.8
#)

options(ucscChromosomeNames=FALSE)
plotTracks(
    #list(itrack, gtrack, altrack, altrack1, altrack2),
    list(gtrack, altrack, altrack1, altrack2),
    type = "coverage",
    chromosome = "chr2", 
    #ucscChromosomeNames=FALSE,
    from = st,
    to = en
)
```

```{r, fig.width=10, fig.height=6}
thechr <- "chr2"
st <- 197405115
en <- 197405145
library(BSgenome.Hsapiens.UCSC.hg38)
strack <- SequenceTrack(
    Hsapiens,
    chromosome = thechr, 
    from = st, 
    to = en,
    cex=1
)

options(ucscChromosomeNames=FALSE)
plotTracks(
    list(gtrack, altrack, altrack1, altrack2, strack), 
    type = "coverage",
    #min.height = 12, max.height = 15, coverageHeight = 0.15, 
    chromosome = "chr2", 
    #complement = TRUE, 
    #ucscChromosomeNames=FALSE,
    from = st,
    to = en
)
```



#make barplot
```{r, fig.width=6, fig.height=5}
fq <- list.files("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/demultiplex/")
fq <- fq[grep("fq.gz", fq)]
sample.ls <- sapply(strsplit(fq, "\\."), function(x){x[1]})

all_info <- data.frame()

for(i in sample.ls){
print(paste("now working on", i, "================================="))
bamPath <- paste0("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/re_basecall/rachseq/fastq_polyA/flames_out/", i, "_align2genome.bam")

mygene <- GRanges(seqnames = "chr2", 
                  ranges = IRanges(197405131, 197405135),
                  strand = "-")
x <- stackStringsFromBam(bamPath, param = mygene, use.names = T)
table(x)
df <- data.frame(read_name = names(as.character(x)), pos = as.character(x))
df$pos_1 <- substring(df$pos, 1, 1) == "T"
df$pos_2 <- substring(df$pos, 3, 3) == "T"
df$pos_3 <- substring(df$pos, 5, 5) == "T"
df$mis_match_score <- rowSums(df[, c("pos_1", "pos_2", "pos_3")])


#select the row with mis-match score > 0 
df2 <- df[df$mis_match_score == 0, ] # all the reads in df are artifacts
tmp <- data.frame(raw = nrow(df), 
                  filtered = nrow(df2), 
                  sample = i)
all_info <- rbind(all_info, tmp)
}
all_info$loss <- all_info$raw - all_info$filtered
all_info <- all_info[,c("sample", "loss", "filtered")]

pt_table <- read.csv("sf3b1_in_t_fig/patient_number.csv", header = F)
colnames(pt_table) <- c("sample", "alias")

all_info$alias <- pt_table$alias[match(all_info$sample, pt_table$sample)]
#alias_order <- df$alias[order(df$nCell, decreasing = T)]
#alias_order_adj <- alias_order[c(1:4, 6:16, 18:20, 17, 5, 21) ]
alias_order_adj <- c("CLL3-S", "CLL3-R", "CLL26-S", "CLL26-R", "CLL2-RB", "CLL2-R", "CLL5-R", "CLL5-RB", "CLL6-S", "CLL6-R", "CLL7-S", "CLL7-R", 
                     "CLL16-S", "CLL16-R", "CLL17-R", "CLL23-S", "CLL23-R", "CLL23-RB",
                     "HD11", "HD12", "HD14")

all_info$alias <- factor(all_info$alias, levels = alias_order_adj)

plot_df <- gather(all_info, category, count, loss:filtered)
plot_df$category <- factor(plot_df$category, levels = c("loss", "filtered"))
ggplot(plot_df, aes(x = alias, y = count, fill = category)) +
  geom_col(position = "identity") +
  theme_bw() +
  scale_fill_manual(values = c("grey", "grey50")) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
ggsave("sf3b1_artefact_fig/barplot_reads_left_after_artifact_removal.pdf", width = 5.5, height = 5)
```


```{r, fig.width=6, fig.height=5}
fq <- list.files("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/demultiplex/")
fq <- fq[grep("fq.gz", fq)]
sample.ls <- sapply(strsplit(fq, "\\."), function(x){x[1]})

all_info <- data.frame()

for(i in sample.ls){
print(paste("now working on", i, "================================="))
bamPath <- paste0("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/re_basecall/rachseq/fastq_polyA/flames_out/", i, "_align2genome.bam")

#read in SF3B1 gene region
mygene <- GRanges(seqnames = "chr2", 
                  ranges = IRanges(197405131, 197405135),
                  strand = "-")
x <- stackStringsFromBam(bamPath, param = mygene, use.names = T)
#table(x)
df <- data.frame(read_name = names(as.character(x)), pos = as.character(x))
df$pos_1 <- substring(df$pos, 1, 1) == "T"
df$pos_2 <- substring(df$pos, 3, 3) == "T"
df$pos_3 <- substring(df$pos, 5, 5) == "T"
df$mis_match_score <- rowSums(df[, c("pos_1", "pos_2", "pos_3")])
df$bc <- sapply(strsplit(df$read_name, "_"), function(x){x[1]})
df$bc_umi <- sapply(strsplit(df$read_name, "#"), function(x){x[1]})

#select the row with mis-match score = 0 
df2 <- df[df$mis_match_score == 0, ]


tmp <- data.frame(raw_cell = length(unique(df$bc)), 
                  raw_umi = length(unique(df$bc_umi)),
                  filtered_cell = length(unique(df2$bc)), 
                  filtered_umi = length(unique(df2$bc_umi)),
                  sample = i)
all_info <- rbind(all_info, tmp)
}

all_info$raw <- all_info$raw_umi / all_info$raw_cell
all_info$filtered <- all_info$filtered_umi / all_info$filtered_cell

backup <- all_info
# plot_df <- gather(all_info, category, count, raw:filtered)
# plot_df$category <- factor(plot_df$category, levels = c("raw", "filtered"))
# ggplot(plot_df, aes(x = sample, y = count, fill = category)) +
#   geom_col(position = "identity") +
#   theme_bw() +
#   scale_fill_manual(values = c("grey", "grey50")) +
#   theme(axis.text.x = element_text(angle = 45, hjust = 1))
# ggsave("QC_step2_figure/barplot_reads_left_after_artifact_removal.pdf", width = 5.5, height = 5)
```


```{r, fig.width=5, fig.height=5}
df_abline <- data.frame(intercept = 0, slope = 1)
df_abline_1 <- data.frame(intercept = 0, slope = 0.5)
ggplot(all_info, aes(x = raw, y = filtered)) + 
  geom_point(color = "#E47250", size = 2.5) +
  theme_classic() +
  xlim(c(0, 15)) + 
  ylim(c(0, 15)) + 
  xlab("Unfiltered") + 
  ylab("Filtered") + 
  geom_abline(data=df_abline, aes(intercept=intercept,slope=slope), linetype = "dashed") +
  theme(axis.text = element_text(size = 15), 
        axis.title = element_text(size = 15)) +
  ggtitle("Per cell SF3B1 UMIs after removing artifacts")
ggsave("sf3b1_artefact_fig/per_cell_umi_after_removing_artifact.pdf", width = 5, height = 5)
```



```{r, fig.width=5, fig.height=5}
sr <- readRDS("make_srt_object/sr_data.rds")
cor.df <- data.frame(barcode = colnames(sr), 
                     sr_sf3b1 = sr@assays$RNA@counts["SF3B1",])
cor.df$sample <- sapply(strsplit(cor.df$barcode, "_"), function(x){x[1]})

sr.df <- table(cor.df$sample) %>% as.data.frame()
colnames(sr.df) <- c("sample", "sr_cell")

cor.df <- cor.df %>% group_by(sample) %>% summarise(sr_umi = sum(sr_sf3b1)) %>% as.data.frame()
sr.df <- full_join(cor.df, sr.df, by = "sample")
sr.df$sr <- sr.df$sr_umi / sr.df$sr_cell

all_info <- full_join(sr.df, all_info, by = "sample")

df_abline <- data.frame(intercept = 0, slope = 1)
ggplot(all_info, aes(x = sr, y = filtered)) + 
  geom_point(color = "#E47250", size = 2.5) +
  theme_classic() +
  xlim(c(0, 5)) + 
  ylim(c(0, 5)) + 
  xlab("Short-read data") + 
  ylab("Filtered long-read data") + 
  ggtitle("Per cell UMIs in filtered long-/short-read data") + 
  geom_abline(data=df_abline, aes(intercept=intercept,slope=slope), linetype = "dashed") +
  theme(axis.text = element_text(size = 15), 
        axis.title.x = element_text(size = 15),
        axis.title.y = element_text(size = 15))
ggsave("sf3b1_artefact_fig/per_cell_umi_filtered_lr_sr.pdf", width = 5, height = 5)
```


```{r, fig.width=6, fig.height=5}
cell_num <- all_info[, c("sample", "raw_cell", "filtered_cell")]
cell_num$cell_loss <- cell_num$raw_cell - cell_num$filtered_cell
cell_num <- cell_num[, c("sample", "filtered_cell", "cell_loss")]

cell_num <- gather(cell_num, Group, Count, filtered_cell:cell_loss)

pt_table <- read.csv("sf3b1_in_t_fig/patient_number.csv", header = F)
colnames(pt_table) <- c("sample", "alias")

cell_num$alias <- pt_table$alias[match(cell_num$sample, pt_table$sample)]
#alias_order <- df$alias[order(df$nCell, decreasing = T)]
#alias_order_adj <- alias_order[c(1:4, 6:16, 18:20, 17, 5, 21) ]
alias_order_adj <- c("CLL3-S", "CLL3-R", "CLL26-S", "CLL26-R", "CLL2-RB", "CLL2-R", "CLL5-R", "CLL5-RB", "CLL6-S", "CLL6-R", "CLL7-S", "CLL7-R", 
                     "CLL16-S", "CLL16-R", "CLL17-R", "CLL23-S", "CLL23-R", "CLL23-RB",
                     "HD11", "HD12", "HD14")

cell_num$alias <- factor(cell_num$alias, levels = alias_order_adj)

ggplot(cell_num, aes(x = alias, y = Count, fill = Group)) + 
  geom_col() +
  scale_fill_manual(values = c("grey80", "grey40")) + 
  scale_y_continuous(breaks = c(0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000)) + 
  xlab(NULL) +
  ylab("Cell number") + 
  ggtitle("Cell counts after artifact removal") + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10),
        axis.text.y = element_text(size = 10))
ggsave("sf3b1_artefact_fig/cell_num_after_artifact_removal.pdf", width = 6, height = 5)
```



```{r}
pt_table <- read.csv("sf3b1_in_t_fig/patient_number.csv", header = F)
colnames(pt_table) <- c("sample", "alias")

#all cells with SF3B1 K700E
selected_mut_mtx <- mut_mtx[mut_mtx$mut > 0, ]
df <- data.frame(table(selected_mut_mtx$patient))
colnames(df) <- c("sample", "nCell")
df$group <- "SF3B1-WT"
df$group[df$sample %in% c("CLL153", "CLL156", "CLL155", "CLL5306")] <- "SF3B1-K700E"
df$group <- factor(df$group, levels = c("SF3B1-K700E", "SF3B1-WT"))

df$alias <- pt_table$alias[match(df$sample, pt_table$sample)]
#alias_order <- df$alias[order(df$nCell, decreasing = T)]
#alias_order_adj <- alias_order[c(1:4, 6:16, 18:20, 17, 5, 21) ]
alias_order_adj <- c("CLL3-S", "CLL3-R", "CLL26-S", "CLL26-R", "CLL2-RB", "CLL2-R", "CLL5-R", "CLL5-RB", "CLL6-S", "CLL6-R", "CLL7-S", "CLL7-R", 
                     "CLL16-S", "CLL16-R", "CLL17-R", "CLL23-S", "CLL23-R", "CLL23-RB",
                     "HD11", "HD12", "HD14")

df$alias <- factor(df$alias, levels = alias_order_adj)
```
























