---
title: "SF3B1 mutations in T cells"
author: "Hongke Peng"
date: "2023-12-21"
output: html_document
---

#load library
```{r}
library(SingleCellExperiment)
library(clusterProfiler)
library(dittoSeq)
library(DropletUtils)
library(dplyr)
library(ggplot2)
library(Seurat)
library(ggpointdensity)
library(harmony)
library(transPlotR)
library(biomaRt)
library(Seurat)
library(SingleCellExperiment)
```

```{r, fig.width=5, fig.height=5}
sr <- readRDS("make_srt_object/sr_data.rds")
colset <- c("#A9A9A9", 
            "#F055A3", "#A56AAC", "#D99ACE", "#DCBFDD", "#77B3DA", "#4CBBD5", "#A5D0F0", "#528DD9", "#7876B1B2","#4DBBD5B2","#00A1D5B2",
            "#9BCA3C", "#91D1C2", "#5FB14E", "#CADB73", "#3B897C")
  
DimPlot(object = sr, reduction = "mnn_umap", order = F, group.by = "refined.cluster", 
        cols = colset, raster = T, pt.size = 0.3,
        label = T, label.box = T, label.color = "white", label.size = 3)
```

```{r}
mut_ls <- list()
ref_ls <- list()
for(i in c("CLL156","CLL5309","CLL5602","CLL5305","CLL5306","CLL108",
           "CLL141v","CLL141i","CLL153","CLL63i","CLL63v","CLL232","CLL152","CLL318","CLL155","CLL170i", "CLL170v", "CLL175",
           "HD11","HD12","HD14")) {
  mut_pathway <- paste0("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/re_basecall/rachseq/fastq_polyA/filtered_bam/", i, "_mutation/alt_cnt.csv.gz") 
  x <- read.csv(mut_pathway)
  rownames(x) <- paste(x$chr, x$position, sep = "_")
  x$chr <- NULL
  x$position <- NULL
  colnames(x) <- paste0(i, "_", colnames(x), "-1")
  keep_cell <- colnames(x)[colnames(x) %in% colnames(sr)]
  x <- x[, keep_cell]
  mut_ls[[i]] <- x
  
  ref_pathway <- paste0("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/re_basecall/rachseq/fastq_polyA/filtered_bam/", i, "_mutation/ref_cnt.csv.gz") 
  x <- read.csv(ref_pathway)
  rownames(x) <- paste(x$chr, x$position, sep = "_")
  x$chr <- NULL
  x$position <- NULL
  colnames(x) <- paste0(i, "_", colnames(x), "-1")
  keep_cell <- colnames(x)[colnames(x) %in% colnames(sr)]
  x <- x[, keep_cell]
  ref_ls[[i]] <- x
}
# sf3b1 mut chr_197402109
```


```{r}
mut_mtx <- data.frame()
for(i in c("CLL156","CLL5309","CLL5602","CLL5305","CLL5306","CLL108",
           "CLL141v","CLL141i","CLL153","CLL63i","CLL63v","CLL232","CLL152","CLL318","CLL155","CLL170i", "CLL170v", "CLL175",
           "HD11","HD12","HD14")) {
  x <- mut_ls[[i]]
  x <- as.data.frame(t(x))
  y <- ref_ls[[i]]
  y <- as.data.frame(t(y))
  
  tmp <- data.frame(bc = rownames(x), 
                   mut = x$chr2_197402109,
                   ref = y$chr2_197402109)
  mut_mtx <- rbind(mut_mtx, tmp)
}
rownames(mut_mtx) <- mut_mtx$bc
mut_mtx <- mut_mtx[rownames(mut_mtx) %in% colnames(sr), ]
mut_mtx$patient <- sapply(strsplit(rownames(mut_mtx), "_"), function(x){x[1]})
```

#check WT samples as controls
* Cell with SF3B1 mutations: should only exist in CLL cells. 
* Only 4 samples have this point mutation
```{r}
mut_mtx$K700E <- mut_mtx$mut > 0
mut_mtx$WT <- mut_mtx$ref > 0
table(mut_mtx$mut, mut_mtx$patient)
table(mut_mtx$ref, mut_mtx$patient)
table(mut_mtx$K700E, mut_mtx$WT)

# I will use all cell with ref > 10 and mut = 0
#ref_mtx <- mut_mtx[mut_mtx$mut == 0 & mut_mtx$ref >= 10, ] # 5961 cells
ref_mtx <- mut_mtx[mut_mtx$ref > 10]
table(ref_mtx$patient)
ref_group <- ref_mtx$bc

# I will use all cell with mut > 0
selected_mut_mtx <- mut_mtx[mut_mtx$mut > 0, ] # 1520 cells
dim(selected_mut_mtx)
table(selected_mut_mtx$patient)
mut_group <- selected_mut_mtx$bc
```


```{r, fig.width=13, fig.height=10/3}
pt_table <- read.csv("sf3b1_in_t_fig/patient_number.csv", header = F)
colnames(pt_table) <- c("sample", "alias")

#all cells with SF3B1 K700E
selected_mut_mtx <- mut_mtx[mut_mtx$mut > 0, ]
df <- data.frame(table(selected_mut_mtx$patient))
colnames(df) <- c("sample", "nCell")
df$group <- "SF3B1-WT"
df$group[df$sample %in% c("CLL153", "CLL156", "CLL155", "CLL5306")] <- "SF3B1-K700E"
df$group <- factor(df$group, levels = c("SF3B1-K700E", "SF3B1-WT"))

df$alias <- pt_table$alias[match(df$sample, pt_table$sample)]
#alias_order <- df$alias[order(df$nCell, decreasing = T)]
#alias_order_adj <- alias_order[c(1:4, 6:16, 18:20, 17, 5, 21) ]
alias_order_adj <- c("CLL3-S", "CLL3-R", "CLL26-S", "CLL26-R", "CLL2-RB", "CLL2-R", "CLL5-R", "CLL5-RB", "CLL6-S", "CLL6-R", "CLL7-S", "CLL7-R", 
                     "CLL16-S", "CLL16-R", "CLL17-R", "CLL23-S", "CLL23-R", "CLL23-RB",
                     "HD11", "HD12", "HD14")

df$alias <- factor(df$alias, levels = alias_order_adj)

p1 <- ggplot(df, aes(x = alias, y = nCell, fill = group)) + 
  geom_col(width = 0.75) + 
  scale_fill_manual(values = c("grey30", "grey")) +
  theme_bw() + 
  ylim(c(0, 1500)) + 
  NoLegend() +
  xlab(NULL) + 
  ggtitle("nCells with > 0 UMIs of mutant SF3B1") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 7))

selected_mut_mtx <- mut_mtx[mut_mtx$mut > 1, ]
df <- data.frame(table(selected_mut_mtx$patient))
colnames(df) <- c("sample", "nCell")
df$group <- "SF3B1-WT"
df$group[df$sample %in% c("CLL153", "CLL156", "CLL155", "CLL5306")] <- "SF3B1-K700E"
df$group <- factor(df$group, levels = c("SF3B1-K700E", "SF3B1-WT"))

df$alias <- pt_table$alias[match(df$sample, pt_table$sample)]
#alias_order <- df$alias[order(df$nCell, decreasing = T)]
df$alias <- factor(df$alias, levels = alias_order_adj)

p2 <- ggplot(df, aes(x = alias, y = nCell, fill = group)) + 
  geom_col(width = 0.75) + 
  scale_fill_manual(values = c("grey30", "grey")) +
  theme_bw() + 
  ylim(c(0, 1500)) +
  NoLegend() +
  ylab(NULL) + 
  xlab(NULL) + 
  ggtitle("nCells with > 1 UMIs of mutant SF3B1") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 7))

selected_mut_mtx <- mut_mtx[mut_mtx$mut > 2, ]
df <- data.frame(table(selected_mut_mtx$patient))
colnames(df) <- c("sample", "nCell")
df$group <- "SF3B1-WT"
df$group[df$sample %in% c("CLL153", "CLL156", "CLL155", "CLL5306")] <- "SF3B1-K700E"
df$group <- factor(df$group, levels = c("SF3B1-K700E", "SF3B1-WT"))

df$alias <- pt_table$alias[match(df$sample, pt_table$sample)]
#alias_order <- df$alias[order(df$nCell, decreasing = T)]
df$alias <- factor(df$alias, levels = alias_order_adj)

p3 <- ggplot(df, aes(x = alias, y = nCell, fill = group)) + 
  geom_col(width = 0.75) + 
  scale_fill_manual(values = c("grey30","grey")) +
  theme_bw() + 
  ylim(c(0, 1500)) +
  ylab(NULL) +
  xlab(NULL) + 
  ggtitle("nCells with > 2 UMIs of mutant SF3B1") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 7))

p <- p1 + p2 + p3
p

ggsave(plot = p, filename = "sf3b1_in_t_fig/ncell_with_mut.pdf", width = 13, height = 10/3)
```


```{r}
#all cells with SF3B1 K700E
selected_mut_mtx <- mut_mtx[mut_mtx$mut > 0, ]
df <- data.frame(table(selected_mut_mtx$patient))
colnames(df) <- c("sample", "nCell")
df$group <- "SF3B1-WT"
df$group[df$sample %in% c("CLL153", "CLL156", "CLL155", "CLL5306")] <- "SF3B1-K700E"
df$group <- factor(df$group, levels = c("SF3B1-K700E", "SF3B1-WT"))

df$alias <- pt_table$alias[match(df$sample, pt_table$sample)]
#df$alias <- factor(df$alias, levels = alias_order)
df$category <- "UMIs > 0"
all_df <- df

selected_mut_mtx <- mut_mtx[mut_mtx$mut > 1, ]
df <- data.frame(table(selected_mut_mtx$patient))
colnames(df) <- c("sample", "nCell")
df$group <- "SF3B1-WT"
df$group[df$sample %in% c("CLL153", "CLL156", "CLL155", "CLL5306")] <- "SF3B1-K700E"
df$group <- factor(df$group, levels = c("SF3B1-K700E", "SF3B1-WT"))

df$alias <- pt_table$alias[match(df$sample, pt_table$sample)]
#df$alias <- factor(df$alias, levels = alias_order)
df$category <- "UMIs > 1"
all_df <- rbind(all_df, df)

selected_mut_mtx <- mut_mtx[mut_mtx$mut > 2, ]
df <- data.frame(table(selected_mut_mtx$patient))
colnames(df) <- c("sample", "nCell")
df$group <- "SF3B1-WT"
df$group[df$sample %in% c("CLL153", "CLL156", "CLL155", "CLL5306")] <- "SF3B1-K700E"
df$group <- factor(df$group, levels = c("SF3B1-K700E", "SF3B1-WT"))

df$alias <- pt_table$alias[match(df$sample, pt_table$sample)]
#df$alias <- factor(df$alias, levels = alias_order)
df$category <- "UMIs > 2"
all_df <- rbind(all_df, df)

alias_order_adj <- alias_order[c(1:4, 6:16, 18:20, 17, 5, 21) ]
all_df$alias <- factor(df$alias, levels = alias_order_adj)
ggplot(all_df, aes(x = alias, y = nCell, fill = category)) + 
  geom_col(width = 0.75, position="identity") + 
  scale_fill_manual(values = c("grey80","grey60","grey20")) +
  theme_classic() + 
  ylim(c(0, 1600)) +
  ylab(NULL) +
  xlab(NULL) + 
  ggtitle("UMI counts of mutant SF3B1") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10))
ggsave("sf3b1_in_t_fig/UMIs_filter.pdf", width = 8, height = 5)
```
#check mutation in T cell
```{r, fig.width=5, fig.height=5}
# I will use all cell with mut > 0
selected_mut_mtx <- mut_mtx[mut_mtx$mut > 0, ] # 5059 cells
dim(selected_mut_mtx)
table(selected_mut_mtx$patient)
mut_group <- selected_mut_mtx$bc

sr$mut <- "background"
sr$mut[colnames(sr) %in% mut_group] <- "Mut"
table(sr$mut)

mut_df <- data.frame(sr@reductions$mnn_umap@cell.embeddings, 
                     mut = sr$mut)

mut_df_1 <- mut_df[mut_df$mut == "Mut", ]
mut_df_1$patient <- sapply(strsplit(rownames(mut_df_1), "_"), function(x){x[1]})
mut_df_1 <- mut_df_1[mut_df_1$patient %in% c("CLL153", "CLL156", "CLL155", "CLL5306"), ]
  
ggplot(mut_df) + 
  geom_point(aes(x = mnnumap_1, y = mnnumap_2), color = "grey", size = 0.1) +
  theme_bw() + 
  geom_point(data = mut_df_1, aes(x = mnnumap_1, y = mnnumap_2), color = "grey50", size = 1) +
  xlab("UMAP_1") +
  ylab("UMAP_2")

ggsave("sf3b1_in_t_fig/umap_cell_with_>0_mut_read.png", width = 5, height = 5)
```


```{r, fig.width=5, fig.height=5}
DimPlot(sr, reduction = "mnn_umap", group.by = "refined.cluster", label = T)
sr$cell.type <- "CLL cells"
sr$cell.type[sr$refined.cluster == 0] <- "Monocytes"
sr$cell.type[sr$refined.cluster == 11] <- "B cells"
sr$cell.type[sr$refined.cluster %in% 12:16] <- "T cells"

sr.pt <- subset(sr, orig.ident %in% c("CLL153", "CLL156", "CLL155", "CLL5306"))

df <- table(sr.pt$cell.type, sr.pt$mut)[, "Mut"] %>% as.data.frame()
colnames(df) <- "nCell"
df$cell.type = rownames(df)

df$cell.type <- factor(df$cell.type, levels = c("CLL cells", "B cells", "T cells", "Monocytes"))
ggplot(df, aes(x = cell.type, y = nCell)) + 
  geom_col() + 
  xlab(NULL) + 
  #ylab(NULL) + 
  theme_classic() + 
  ggtitle("Cells with SF3B1 mut reads in subsets") +
  theme(axis.text.x = element_text(size = 15))
  
ggsave("sf3b1_in_t_fig/mut_cell_in_cluster.pdf", width = 5, height = 5)
```

```{r, fig.width=10, fig.height=5}
table(sr$cell.type, sr$mut)
sr.mut <- subset(sr.pt, mut == "Mut")
sr.mut$cell.type <- factor(sr.mut$cell.type, levels = c("CLL cells", "B cells", "T cells", "Monocytes"))
DefaultAssay(sr.mut) <- "RNA"

VlnPlot(sr.mut, group.by = "cell.type", features = c("nCount_RNA", "nFeature_RNA", "percent.mt", "CD3E", "CD14", "CD19"), pt.size = 0, 
        cols = c("#BC3C2999", "#0072B599", "#0072B599", "#0072B599"))

ggsave("sf3b1_in_t_fig/vlnplot_mut_cell_in_T.pdf", width = 10, height = 5)
```

```{r, fig.width=6, fig.height=3}
selected_mut_mtx <- mut_mtx[mut_mtx$patient %in% c("CLL153", "CLL156", "CLL155", "CLL5306"), ]
selected_mut_mtx <- selected_mut_mtx[selected_mut_mtx$mut > 0, ]

CLL_name <- colnames(sr)[sr$cell.type == "CLL cells"]
#other_name <- colnames(sr)[sr$cell.type != "CLL cells"]
selected_mut_mtx$cell.type <- "non-CLL"
selected_mut_mtx$cell.type[selected_mut_mtx$bc %in% CLL_name] <- "CLL"
selected_mut_mtx$cell.type <- factor(selected_mut_mtx$cell.type, levels = c("CLL", "non-CLL"))

ggplot(selected_mut_mtx, aes(x = mut, fill = cell.type)) + 
  geom_bar(alpha = 0.5, binwidth = 1, width = 0.8, position = "identity") + 
  xlim(c(0,50)) + 
  scale_fill_manual(values = ggsci::pal_nejm(alpha = 1)(2)) + 
  geom_vline(xintercept = 2.5, linetype = "dashed") +
  ggtitle("Distribution of UMIs of mutant SF3B1 in cells") + 
  xlab("UMI counts of mutant SF3B1") + 
  ylab("Cell counts") + 
  theme_bw()
ggsave("sf3b1_in_t_fig/read_distr.pdf", width = 6, height = 3)
```


# calculate sequencing error in K700E
```{r, fig.width=8, fig.height=5}
fq <- list.files("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/demultiplex/")
fq <- fq[grep("fq.gz", fq)]
sample.ls <- sapply(strsplit(fq, "\\."), function(x){x[1]})

all_sample <- data.frame()
library(GenomicAlignments)
for (i in sample.ls) {
  print(paste("now working on", i, "==================="))
  bamPath <- paste0("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/re_basecall/rachseq/fastq_polyA/filtered_bam/" , i, "_align2genome.bam")
  mygene <- GRanges(seqnames = "chr2", 
                  ranges = IRanges(197402110, 197402110),
                  strand = "-")
  x <- stackStringsFromBam(bamPath, param = mygene, use.names = T)
  x <- table(x) %>% as.data.frame()
  x <- x[1:4, ]
  colnames(x) <- c("N", "counts")
  x$fraction <- x$counts / sum(x$counts)
  x$sample <- i
  all_sample <- rbind(all_sample, x)
}


all_sample$N <- factor(all_sample$N, levels = c("T", "C", "G", "A"))
all_sample$alias <- pt_table$alias[match(all_sample$sample, pt_table$sample)]
alias_order_adj <- c("CLL3-S", "CLL3-R", "CLL26-S", "CLL26-R", "CLL2-RB", "CLL2-R", "CLL5-R", "CLL5-RB", "CLL6-S", "CLL6-R", "CLL7-S", "CLL7-R", 
                     "CLL16-S", "CLL16-R", "CLL17-R", "CLL23-S", "CLL23-R", "CLL23-RB",
                     "HD11", "HD12", "HD14")
all_sample$alias <- factor(all_sample$alias, levels = alias_order_adj)
ggplot(all_sample, aes(x = alias, y = fraction, fill = N)) + 
  geom_col(width = 0.7) + 
  scale_fill_manual(values = c("grey", ggsci::pal_npg(alpha = 0.8)(3))) +
  theme_classic() + 
  geom_hline(yintercept = 0.02, linetype = "dashed") +
  xlab(NULL) +
  ylab(NULL) + 
  ggtitle("UMIs of reads with Ch2:197402110, T to C alterations") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        axis.text.y = element_text(hjust = 1, size = 12))
ggsave("sf3b1_in_t_fig/k700e_seq_error.pdf", width = 8, height = 5)
``` 
#check orther mutations
```{r, fig.width=8, fig.height=5}
fq <- list.files("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/demultiplex/")
fq <- fq[grep("fq.gz", fq)]
sample.ls <- sapply(strsplit(fq, "\\."), function(x){x[1]})

all_sample <- data.frame()
library(GenomicAlignments)
for (i in sample.ls) {
  print(paste("now working on", i, "==================="))
  bamPath <- paste0("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/re_basecall/rachseq/fastq_polyA/filtered_bam/" , i, "_align2genome.bam")
  mygene <- GRanges(seqnames = "chr2", 
                  ranges = IRanges(197392284, 197392284),
                  strand = "-")
  x <- stackStringsFromBam(bamPath, param = mygene, use.names = T)
  x <- table(x) %>% as.data.frame()
  x <- x[1:4, ]
  colnames(x) <- c("N", "counts")
  x$fraction <- x$counts / sum(x$counts)
  x$sample <- i
  all_sample <- rbind(all_sample, x)
}


all_sample$N <- factor(all_sample$N, levels = c("C", "G", "A", "T"))
all_sample$alias <- pt_table$alias[match(all_sample$sample, pt_table$sample)]
alias_order_adj <- c("CLL3-S", "CLL3-R", "CLL26-S", "CLL26-R", "CLL2-RB", "CLL2-R", "CLL5-R", "CLL5-RB", "CLL6-S", "CLL6-R", "CLL7-S", "CLL7-R", 
                     "CLL16-S", "CLL16-R", "CLL17-R", "CLL23-S", "CLL23-R", "CLL23-RB",
                     "HD11", "HD12", "HD14")
all_sample$alias <- factor(all_sample$alias, levels = alias_order_adj)
ggplot(all_sample, aes(x = alias, y = fraction, fill = N)) + 
  geom_col(width = 0.7) + 
  scale_fill_manual(values = c("grey", ggsci::pal_npg(alpha = 0.8)(3))) +
  theme_classic() + 
  geom_hline(yintercept = 0.02, linetype = "dashed") +
  xlab(NULL) +
  ylab(NULL) + 
  ggtitle("UMIs of reads with Ch2r:197392284, C to G alteration") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        axis.text.y = element_text(hjust = 1, size = 12))
ggsave("sf3b1_in_t_fig/chr2_197392284_seq_error.pdf", width = 8, height = 5)
``` 

## mapping this mutation back to umap
###chr2_197392283
```{r}
mut_mtx <- data.frame()
for(i in c("CLL156","CLL153","CLL63i","CLL63v")) {
  x <- mut_ls[[i]]
  x <- as.data.frame(t(x))
  y <- ref_ls[[i]]
  y <- as.data.frame(t(y))
  
  tmp <- data.frame(bc = rownames(x), 
                   mut = x$chr2_197392283,
                   ref = y$chr2_197392283)
  mut_mtx <- rbind(mut_mtx, tmp)
}
rownames(mut_mtx) <- mut_mtx$bc
mut_mtx <- mut_mtx[rownames(mut_mtx) %in% colnames(sr), ]
mut_mtx$patient <- sapply(strsplit(rownames(mut_mtx), "_"), function(x){x[1]})
```

```{r, fig.width=5, fig.height=5}
# I will use all cell with mut > 2
selected_mut_mtx <- mut_mtx[mut_mtx$mut > 2, ] # 1520 cells
dim(selected_mut_mtx)
table(selected_mut_mtx$patient)
mut_group <- selected_mut_mtx$bc

sr$mut <- "background"
sr$mut[colnames(sr) %in% mut_group] <- "Mut"
table(sr$mut)

mut_df <- data.frame(sr@reductions$mnn_umap@cell.embeddings, 
                     mut = sr$mut)

mut_df_1 <- mut_df[mut_df$mut == "Mut", ]
mut_df_1$patient <- sapply(strsplit(rownames(mut_df_1), "_"), function(x){x[1]})
#mut_df_1 <- mut_df_1[mut_df_1$patient %in% c("CLL156","CLL153","CLL63i","CLL63v"), ]
  
ggplot(mut_df) + 
  geom_point(aes(x = mnnumap_1, y = mnnumap_2), color = "grey", size = 0.1) +
  theme_bw() + 
  geom_point(data = mut_df_1, aes(x = mnnumap_1, y = mnnumap_2), color = "grey50", size = 1) +
  xlab("UMAP 1") + 
  ylab("UMAP 2")

ggsave("sf3b1_in_t_fig/umap_chr2_197392284>2_mut_read.png", width = 5, height = 5)
```

#check chr2 197421096

```{r, fig.width=8, fig.height=5}
fq <- list.files("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/demultiplex/")
fq <- fq[grep("fq.gz", fq)]
sample.ls <- sapply(strsplit(fq, "\\."), function(x){x[1]})

all_sample <- data.frame()
library(GenomicAlignments)
for (i in sample.ls) {
  print(paste("now working on", i, "==================="))
  bamPath <- paste0("/stornext/Genomics/data/CLL_venetoclax/RaCHseq_paper/re_basecall/rachseq/fastq_polyA/filtered_bam/" , i, "_align2genome.bam")
  mygene <- GRanges(seqnames = "chr2", 
                  ranges = IRanges(197421097, 197421097),
                  strand = "-")
  x <- stackStringsFromBam(bamPath, param = mygene, use.names = T)
  x <- table(x) %>% as.data.frame()
  x <- x[1:4, ]
  colnames(x) <- c("N", "counts")
  x$fraction <- x$counts / sum(x$counts)
  x$sample <- i
  all_sample <- rbind(all_sample, x)
}


all_sample$N <- factor(all_sample$N, levels = c("C", "T", "A", "G"))
all_sample$alias <- pt_table$alias[match(all_sample$sample, pt_table$sample)]
alias_order_adj <- c("CLL3-S", "CLL3-R", "CLL26-S", "CLL26-R", "CLL2-RB", "CLL2-R", "CLL5-R", "CLL5-RB", "CLL6-S", "CLL6-R", "CLL7-S", "CLL7-R", 
                     "CLL16-S", "CLL16-R", "CLL17-R", "CLL23-S", "CLL23-R", "CLL23-RB",
                     "HD11", "HD12", "HD14")
all_sample$alias <- factor(all_sample$alias, levels = alias_order_adj)
ggplot(all_sample, aes(x = alias, y = fraction, fill = N)) + 
  geom_col(width = 0.7) + 
  scale_fill_manual(values = c("grey", ggsci::pal_npg(alpha = 0.8)(3))) +
  theme_classic() + 
  geom_hline(yintercept = 0.02, linetype = "dashed") +
  xlab(NULL) +
  ylab(NULL) + 
  ggtitle("UMIs of reads with Ch2r:197421097, C to T alteration") + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        axis.text.y = element_text(hjust = 1, size = 12))
ggsave("sf3b1_in_t_fig/chr2_197421097_seq_error.pdf", width = 8, height = 5)
``` 


```{r}
mut_mtx <- data.frame()
for(i in names(mut_ls)) {
  x <- mut_ls[[i]]
  x <- as.data.frame(t(x))
  y <- ref_ls[[i]]
  y <- as.data.frame(t(y))
  
  tmp <- data.frame(bc = rownames(x), 
                   mut = x$chr2_197421096,
                   ref = y$chr2_197421096)
  mut_mtx <- rbind(mut_mtx, tmp)
}
rownames(mut_mtx) <- mut_mtx$bc
mut_mtx <- mut_mtx[rownames(mut_mtx) %in% colnames(sr), ]
mut_mtx$patient <- sapply(strsplit(rownames(mut_mtx), "_"), function(x){x[1]})
```


```{r, fig.width=5, fig.height=5}
# I will use all cell with mut > 0
selected_mut_mtx <- mut_mtx[mut_mtx$mut > 2, ] # 1520 cells
dim(selected_mut_mtx)
table(selected_mut_mtx$patient)
mut_group <- selected_mut_mtx$bc

sr$mut <- "background"
sr$mut[colnames(sr) %in% mut_group] <- "Mut"
table(sr$mut)

mut_df <- data.frame(sr@reductions$mnn_umap@cell.embeddings, 
                     mut = sr$mut)

mut_df_1 <- mut_df[mut_df$mut == "Mut", ]
mut_df_1$patient <- sapply(strsplit(rownames(mut_df_1), "_"), function(x){x[1]})
#mut_df_1 <- mut_df_1[mut_df_1$patient %in% c("CLL156","CLL153","CLL63i","CLL63v"), ]
  
ggplot(mut_df) + 
  geom_point(aes(x = mnnumap_1, y = mnnumap_2), color = "grey", size = 0.1) +
  theme_bw() + 
  geom_point(data = mut_df_1, aes(x = mnnumap_1, y = mnnumap_2), color = "grey50", size = 1) +
  xlab("UMAP 1") + 
  ylab("UMAP 2")

ggsave("sf3b1_in_t_fig/umap_chr2_197421097>2_mut_read.png", width = 5, height = 5)
```

```{r}
mut_mtx$group <- "wt"
mut_mtx$group[mut_mtx$patient %in% c("CLL156","CLL5306","CLL153","CLL155")] <- "mut"
# I will use all cell with mut > 0
selected_mut_mtx <- mut_mtx[mut_mtx$mut > 1, ] # 1520 cells
dim(selected_mut_mtx)
table(selected_mut_mtx$patient)
mut_group <- selected_mut_mtx$bc
```







#end 